Add minimal modal run to model dry deposition

zdaq12 opened the pull request.

zdaq12 committed:
adding scaffolding for modal run

zdaq12 committed:
add modal run to driver

zdaq12 committed:
add run_modal.F90 to CMakeLists

zdaq12 committed:
add input and output routines for modal run

zdaq12 committed:
add input and output routines for modal parameters

zdaq12 committed:
fix use statements and typos

zdaq12 committed:
fix typos

zdaq12 committed:
add subroutines for computing integrated deposition velocities for the modal run; handle input and output for processing

zdaq12 committed:
add support for QUADPACK; comment out dry dep scenario for now

zdaq12 committed:
remove debugging code for modal run

zdaq12 committed:
clean up implementation of numerical integration with QUADPACK

zdaq12 committed:
formatting

zdaq12 committed:
remove unused flag for emissions/dilution

zdaq12 committed:
fix conditional for inputting dry deposition parameterization

zdaq12 committed:
allow for reading drydep_params from text file otherwise fall back on default parameters

zdaq12 committed:
remove debugging and commented code

zdaq12 committed:
update MPI routines to handle drydep parameters

zdaq12 committed:
update processing script for modal runs and add exectuable

zdaq12 committed:
update aero_dist netcdf input and output for modal runs to operate with arrays

zdaq12 committed:
remove duplicated line

zdaq12 committed:
update dockerignore

zdaq12 committed:
dry_dep --> drydep

zdaq12 committed:
add input files and scripts for modal drydep runs

zdaq12 committed:
update doc for spec_file_read_drydep_params

zdaq12 committed:
fix aero_dist naming convention

zdaq12 committed:
drydep_params.dat comments update

zdaq12 committed:
text file clean up

zdaq12 committed:
newlines at end of files

zdaq12 committed:
minor refactor

zdaq12 committed:
Merge branch 'master' into final_modal

jcurtis2 commented on src/scenario.F90:
```diff
@@ -506,36 +603,308 @@ real(kind=dp) function scenario_loss_rate_dry_dep(vol, density, aero_data, &
+
+#ifdef PMC_USE_QUADPACK
+     ! Do numerical integration when QUADPACK is available
+     print *, "***USING QUADPACK FOR DRY DEPOSITION***"
```
I don't think we need this - will print frequently.

```suggestion
```

jcurtis2 commented on src/scenario.F90:
```diff
@@ -506,36 +603,308 @@ real(kind=dp) function scenario_loss_rate_dry_dep(vol, density, aero_data, &
+    if (ier > 0) then
+     call die_msg(837465, "QUADPACK integration failed (error code: " &
+                         // trim(integer_to_string(ier)))
+    endif
```
Suggest that we switch to `assert_msg`. Test for ier = zero rather than ier > 0 (unless there is a good reason for testing >0 which would include negatives (if quadpack allows it?)). And replace number with random 9 digit number. Suggestion would look somewhat like the following (admittedly I didn't test it yet)

```suggestion
call assert_msg(909106718, ier == 0, &
     "QUADPACK integration failed, error code: " &
     // trim(integer_to_string(ier)))
```

jcurtis2 commented on src/scenario.F90:
```diff
@@ -392,6 +425,77 @@ subroutine scenario_update_aero_binned(scenario, delta_t, env_state, &
+          aero_dist%mode(i_mode)%num_conc = 0d0
+          aero_dist%mode(i_mode)%char_radius = 0d0
+          cycle
+       end if
```
Is this some sort of cutoff for small number concentrations? I think a comment would help.

And is it necessary to set the `char_radius` to 0 or is num_conc = 0 sufficient? I can envision this could cause some issue (maybe currently only in post-processing) if char_radius = 0 is ever used in a lognormal equation.

jcurtis2 commented on src/scenario.F90:
```diff
@@ -392,6 +425,77 @@ subroutine scenario_update_aero_binned(scenario, delta_t, env_state, &
+    else if (scenario%loss_function_type == SCENARIO_LOSS_FUNCTION_CONSTANT) then
+       return
+    else if (scenario%loss_function_type == SCENARIO_LOSS_FUNCTION_DRYDEP) then
+      ! loss
```
```suggestion
```

jcurtis2 commented on src/scenario.F90:
```diff
@@ -392,6 +425,77 @@ subroutine scenario_update_aero_binned(scenario, delta_t, env_state, &
+    !> Environment state.
+    type(env_state_t), intent(in) :: env_state
+    !> Density for each mode.
+    real(kind=dp), intent(in) :: density
```
Comment says each mode but density is a scalar.

jcurtis2 commented on src/scenario.F90:
```diff
@@ -1000,6 +1380,115 @@ end subroutine spec_file_read_scenario
+    !> Dry deposition parameters.
+    type(drydep_params_t), intent(inout) :: drydep_params
+
+    !> \page input_format_chamber Input File Format: Dry Deposition Parameters
```
Copy/pasted \page from elsewhere (chamber.F90).  Should be renamed.

jcurtis2 commented on src/scenario.F90:
```diff
@@ -392,6 +425,77 @@ subroutine scenario_update_aero_binned(scenario, delta_t, env_state, &
+       ! New geometric mean diameter
+       new_d_pg = (new_M / new_N * exp(-(3.0d0**2.0d0)/2.0d0 * (ln_sigma_g**2.0d0)))**(1.0d0/3.0d0)
+       aero_dist%mode(i_mode)%char_radius = new_d_pg / 2.0d0
+      end do
```
Fix indentation on the do loop.

jcurtis2 commented on scenarios/7_drydep/drydep_modal.spec:
```diff
@@ -0,0 +1,33 @@
+run_type modal                       # modal run
+output_prefix data/modal_dpg_00001000000000000000_sig_2_5_emerson_grass
```
Suggest that you change the directory to out (consistent with other scenarios and also specified in the dockerignore)

jcurtis2 commented on scenarios/7_drydep/2_process_modal.sh:
```diff
@@ -0,0 +1,10 @@
+
+# The data should have already been generated by ./1_run_dry_dep_modal.sh
+
+../../build/drydep_modal_process
```
Complains that it needs an output prefix.

jcurtis2 commented on scenarios/7_drydep/aero_data.dat:
```diff
@@ -0,0 +1,2 @@
+#     dens (kg/m^3)   ions in soln (1)    molec wght (kg/mole)   kappa (1)
+SO4            1800                0                   96d-3      0.65
```
Need to tack on the two new freezing parameters. Should be something like this:

```suggestion
#     dens (kg/m^3)   ions in soln (1)    molec wght (kg/mole)   kappa (1)  abifm_m (1)      abifm_c (1)
SO4            1800                0                   96d-3      0.65    0.       0.
```

jcurtis2 reviewed (commented):
Looks pretty good in my first pass.

- I think we need to add some protections to the modal model. For example, I can attempt to run with emissions/dilution (it won't complain) but they won't do anything. However these are in the spec file (they're zero for the purpose of this scenario but it leads one to think that maybe they could be changed) and they are read in (due to other PartMC subroutines). I think maybe we just add some asserts in partmc_modal to warn/error out on unsupported behavior.
- Similarly, I could also attempt to turn coagulation on. This isn't actually read anywhere. I guess what we really need is an input_format_modal similar to how we have them for input_format_particle, sectional to say what is supported.
- We do need to figure out what to do with the scenario. Honestly I'm wondering if we should consider switching this scenario to being a test case (this is what we did with the immersion freezing). This scenario already fits in pretty well with what is found in `test/loss`. Making it a test would also make codecov happy. Alternatively, I think if it wants to reach scenario levels by making it interesting, we could consider running it for the particle-resolved case and/or sectional case. Just something to consider.

jcurtis2 commented on src/scenario.F90:
```diff
@@ -1185,6 +1677,100 @@ subroutine pmc_mpi_unpack_scenario(buffer, position, val)
+    type(drydep_params_t), intent(in) :: val
+
+    integer :: total_size, i, N
+
```
Looks like these variables tagged along from elsewhere and aren't needed here (or actually even where you probably got it from). But clean them up here only so this PR stays focused on drydep changes.

```suggestion
```

zdaq12 committed:
Update src/scenario.F90

Co-authored-by: Jeffrey Curtis <jcurtis2@illinois.edu>

zdaq12 committed:
Update src/scenario.F90

Co-authored-by: Jeffrey Curtis <jcurtis2@illinois.edu>

zdaq12 committed:
Update scenarios/7_drydep/aero_data.dat

Co-authored-by: Jeffrey Curtis <jcurtis2@illinois.edu>

zdaq12 committed:
Update src/scenario.F90

Co-authored-by: Jeffrey Curtis <jcurtis2@illinois.edu>

zdaq12 committed:
address comments

zdaq12 committed:
add missing newline

zdaq12 committed:
guard modal inputs; fix typos; clarity improvements

zdaq12 committed:
add tests for modal dry deposition

zdaq12 committed:
bump codecov version

zdaq12 committed:
fix indentation

zdaq12 committed:
check execution of modal processing script in tests

jcurtis2 commented on scenarios/7_drydep/2_process_modal.sh:
```diff
@@ -0,0 +1,10 @@
+# turn on command echoing
+set -v
+
+# The data should have already been generated by ./1_run_dry_dep_modal.sh
```
```suggestion
# The data should have already been generated by ./1_run_modal.sh
```

jcurtis2 commented on scenarios/7_drydep/1_run_modal.sh:
```diff
@@ -0,0 +1,12 @@
+
+../../build/partmc drydep_modal.spec
+
+# Now run ./2_process_drydep_modal.sh to process the data
```
```suggestion
# Now run ./2_process_modal.sh to process the data
```

jcurtis2 commented on src/aero_dist.F90:
```diff
@@ -271,6 +271,82 @@ subroutine spec_file_read_aero_dist(file, aero_data, &
+    integer, intent(in) :: ncid
+
+    integer :: i_mode, n_mode
+    real(kind=dp), allocatable :: char_rads(:), log10_std_dev_rads(:), num_concs(:)
```
```suggestion
    real(kind=dp), allocatable :: char_rads(:), log10_std_dev_rads(:), &
         num_concs(:)
```

jcurtis2 commented on src/aero_dist.F90:
```diff
@@ -271,6 +271,82 @@ subroutine spec_file_read_aero_dist(file, aero_data, &
+
+    integer :: i_mode, n_mode
+    type(aero_mode_t), allocatable :: modes(:)
+    real(kind=dp), allocatable :: char_rads(:), log10_std_dev_rads(:), num_concs(:)
```
```suggestion
    real(kind=dp), allocatable :: char_rads(:), log10_std_dev_rads(:), &
         num_concs(:)
```

jcurtis2 commented on src/aero_dist.F90:
```diff
@@ -271,6 +271,82 @@ subroutine spec_file_read_aero_dist(file, aero_data, &
+                             unit="1")
+    call pmc_nc_write_real_1d(ncid, num_concs, "num_conc_modes", dim_name="num_modes", &
+                              long_name="number concentration for each mode", &
+                              unit="m^-3")
```
Rearrange a bit to fix long lines (L297, 300, 303)

jcurtis2 commented on src/scenario.F90:
```diff
@@ -1000,6 +1366,115 @@ end subroutine spec_file_read_scenario
+
+!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+
+
```
```suggestion
```

jcurtis2 commented on src/scenario.F90:
```diff
@@ -1000,6 +1366,115 @@ end subroutine spec_file_read_scenario
+    call spec_file_read_real(file, "nu", drydep_params%nu)
+    call spec_file_read_real(file, "beta", drydep_params%beta)
+
+   end subroutine spec_file_read_drydep_params
```
```suggestion
  end subroutine spec_file_read_drydep_params
```

jcurtis2 commented on src/scenario.F90:
```diff
@@ -506,36 +592,305 @@ real(kind=dp) function scenario_loss_rate_dry_dep(vol, density, aero_data, &
+    !> Velocities.
+    real(kind=dp), intent(inout) :: velocities(:)
+
+    integer :: i_mode, n_mode
```
```suggestion
    integer :: i_mode
```

jcurtis2 commented on src/scenario.F90:
```diff
@@ -506,36 +592,305 @@ real(kind=dp) function scenario_loss_rate_dry_dep(vol, density, aero_data, &
+    do i_mode = 1,size(velocities)
+       velocities(i_mode) = scenario_integrated_loss_rate_drydep( &
+          scenario, aero_dist%mode(i_mode), moment, density, env_state) &
+          * env_state%height
```
```suggestion
       velocities(i_mode) = scenario_integrated_loss_rate_drydep( &
            scenario, aero_dist%mode(i_mode), moment, density, env_state) &
            * env_state%height
```

jcurtis2 commented on src/scenario.F90:
```diff
@@ -1185,6 +1663,98 @@ subroutine pmc_mpi_unpack_scenario(buffer, position, val)
+    type(drydep_params_t), intent(inout) :: val
+
+#ifdef PMC_USE_MPI
+      integer :: prev_position
```
```suggestion
    integer :: prev_position
```

jcurtis2 commented on src/scenario.F90:
```diff
@@ -1000,6 +1366,115 @@ end subroutine spec_file_read_scenario
+    call pmc_nc_write_real(ncid, drydep_params%C_IN,    "drydep_C_IN",    unit="1")
+    call pmc_nc_write_real(ncid, drydep_params%C_IM,    "drydep_C_IM",    unit="1")
+    call pmc_nc_write_real(ncid, drydep_params%nu,      "drydep_nu",      unit="1")
+    call pmc_nc_write_real(ncid, drydep_params%beta,    "drydep_beta",    unit="1")
```
To avoid long lines, I recommend just refactoring this into an `associate` block

```
    associate (d => drydep_params)
       call pmc_nc_write_real(ncid, d%z_ref, "drydep_z_ref", unit="m")
       .
       .
       .
    end associate
```

The unpack subroutine does not need it as it doesn't have the units but you could also do it there for visual consistency.

jcurtis2 commented on src/scenario.F90:
```diff
@@ -506,36 +592,305 @@ real(kind=dp) function scenario_loss_rate_dry_dep(vol, density, aero_data, &
+
+#ifdef PMC_USE_QUADPACK
+  real(kind=dp) function scenario_integrated_loss_rate_drydep_quadpack(scenario, &
+      aero_mode, moment, density, env_state)
```
Attempt to address the number of too long lines here to address in the QUADPACK subroutines.

```suggestion
  real(kind=dp) function scenario_integrated_loss_rate_drydep_quadpack( &
      scenario, aero_mode, moment, density, env_state)
```

jcurtis2 commented on src/scenario.F90:
```diff
@@ -506,36 +592,305 @@ real(kind=dp) function scenario_loss_rate_dry_dep(vol, density, aero_data, &
+#ifdef PMC_USE_QUADPACK
+     ! Do numerical integration when QUADPACK is available
+     scenario_integrated_loss_rate_drydep = scenario_integrated_loss_rate_drydep_quadpack( &
+                                              scenario, aero_mode, moment, density, env_state)
```
```suggestion
     scenario_integrated_loss_rate_drydep = &
          scenario_integrated_loss_rate_drydep_quadpack( &
          scenario, aero_mode, moment, density, env_state)
```

jcurtis2 commented on src/scenario.F90:
```diff
@@ -506,36 +592,305 @@ real(kind=dp) function scenario_loss_rate_dry_dep(vol, density, aero_data, &
+    ln_sigma_g = aero_mode%log10_std_dev_radius / log10(exp(1.0d0))
+    ! density of air
+    density_air = (const%air_molec_weight * env_state%pressure) &
+                / (const%univ_gas_const * env_state%temp)
```
```suggestion
    density_air = (const%air_molec_weight * env_state%pressure) &
          / (const%univ_gas_const * env_state%temp)
```

jcurtis2 commented on src/scenario.F90:
```diff
@@ -506,36 +592,305 @@ real(kind=dp) function scenario_loss_rate_dry_dep(vol, density, aero_data, &
+
+    ! Aerodynamic resistance (assuming neutral stability)
+    u_star = 0.4d0 * drydep_params%u_mean / log(drydep_params%z_ref / drydep_params%z_rough)
+    R_a = (1.0d0 / (0.4d0 * u_star)) * log(drydep_params%z_ref / drydep_params%z_rough)
```
```suggestion
    u_star = 0.4d0 * drydep_params%u_mean / log(drydep_params%z_ref &
         / drydep_params%z_rough)
    R_a = (1.0d0 / (0.4d0 * u_star)) * log(drydep_params%z_ref &
         / drydep_params%z_rough)
```

jcurtis2 commented on src/scenario.F90:
```diff
@@ -506,36 +592,305 @@ real(kind=dp) function scenario_loss_rate_dry_dep(vol, density, aero_data, &
+
+    ! Brownian diffusivity
+    D_bar = (const%boltzmann * env_state%temp) / &
+            (3.0d0 * const%pi * visc_d * d_pg)
```
```suggestion
    D_bar = (const%boltzmann * env_state%temp) &
            / (3.0d0 * const%pi * visc_d * d_pg)
```

jcurtis2 commented on src/scenario.F90:
```diff
@@ -506,36 +592,305 @@ real(kind=dp) function scenario_loss_rate_dry_dep(vol, density, aero_data, &
+    ! Loss rate
+    scenario_integrated_loss_rate_drydep = V_d_hat / env_state%height
+
+  end function scenario_integrated_loss_rate_drydep
```
Some slightly more complicated line break and indentation issues remain.

jcurtis2 commented on src/scenario.F90:
```diff
@@ -506,36 +592,305 @@ real(kind=dp) function scenario_loss_rate_dry_dep(vol, density, aero_data, &
+    ln_sigma_g = aero_mode%log10_std_dev_radius / log10(exp(1.0d0))
+    ! density of air
+    density_air = (const%air_molec_weight * env_state%pressure) &
+                / (const%univ_gas_const * env_state%temp)
```
```suggestion
    density_air = (const%air_molec_weight * env_state%pressure) &
         / (const%univ_gas_const * env_state%temp)
```

jcurtis2 commented on src/scenario.F90:
```diff
@@ -506,36 +592,305 @@ real(kind=dp) function scenario_loss_rate_dry_dep(vol, density, aero_data, &
+    real(kind=dp), intent(in) :: density
+    !> Environment state.
+    type(env_state_t), intent(in) :: env_state
+    !> Velocities.
```
Suggest to improve the comment.

jcurtis2 commented on src/scenario.F90:
```diff
@@ -506,36 +592,305 @@ real(kind=dp) function scenario_loss_rate_dry_dep(vol, density, aero_data, &
+    type(scenario_t), intent(in) :: scenario
+    !> Aerosol distribution.
+    type(aero_dist_t), intent(in) :: aero_dist
+    !> Moment to compute rate for.
```
Suggest comment change away from saying rate.

jcurtis2 commented on src/scenario.F90:
```diff
@@ -392,6 +425,66 @@ subroutine scenario_update_aero_binned(scenario, delta_t, env_state, &
+
+  !> Update the modal aerosol distribution to account for particle loss.
+  subroutine scenario_update_aero_modes(aero_dist, del_t, env_state, &
+                                        density, scenario)
```
```suggestion
  subroutine scenario_update_aero_modes(aero_dist, del_t, env_state, &
       density, scenario)
```

jcurtis2 commented on src/output.F90:
```diff
@@ -782,6 +783,149 @@ subroutine input_sectional(filename, index, time, del_t, uuid, bin_grid, &
+  !> Input modal data.
+  subroutine input_modal(filename, index, time, del_t, uuid, aero_dist, &
+                         aero_binned, aero_data, env_state, gas_data, &
+                         gas_state, bin_grid, scenario)
```
```suggestion
  subroutine input_modal(filename, index, time, del_t, uuid, aero_dist, &
       aero_binned, aero_data, env_state, gas_data, gas_state, bin_grid, &
       scenario)
```

zdaq12 committed:
Update scenarios/7_drydep/2_process_modal.sh

Co-authored-by: Jeffrey Curtis <jcurtis2@illinois.edu>

zdaq12 committed:
Update scenarios/7_drydep/1_run_modal.sh

Co-authored-by: Jeffrey Curtis <jcurtis2@illinois.edu>

zdaq12 committed:
Update src/aero_dist.F90

Co-authored-by: Jeffrey Curtis <jcurtis2@illinois.edu>

zdaq12 committed:
Update src/aero_dist.F90

Co-authored-by: Jeffrey Curtis <jcurtis2@illinois.edu>

zdaq12 committed:
Update src/scenario.F90

Co-authored-by: Jeffrey Curtis <jcurtis2@illinois.edu>

zdaq12 committed:
Update src/scenario.F90

Co-authored-by: Jeffrey Curtis <jcurtis2@illinois.edu>

zdaq12 committed:
Update src/scenario.F90

Co-authored-by: Jeffrey Curtis <jcurtis2@illinois.edu>

zdaq12 committed:
Update src/scenario.F90

Co-authored-by: Jeffrey Curtis <jcurtis2@illinois.edu>

zdaq12 committed:
Update src/output.F90

Co-authored-by: Jeffrey Curtis <jcurtis2@illinois.edu>

zdaq12 committed:
Update src/scenario.F90

Co-authored-by: Jeffrey Curtis <jcurtis2@illinois.edu>

zdaq12 committed:
Update src/scenario.F90

Co-authored-by: Jeffrey Curtis <jcurtis2@illinois.edu>

zdaq12 committed:
Update src/scenario.F90

Co-authored-by: Jeffrey Curtis <jcurtis2@illinois.edu>

zdaq12 committed:
Update src/scenario.F90

Co-authored-by: Jeffrey Curtis <jcurtis2@illinois.edu>

zdaq12 committed:
Update src/scenario.F90

Co-authored-by: Jeffrey Curtis <jcurtis2@illinois.edu>

zdaq12 committed:
Update src/scenario.F90

Co-authored-by: Jeffrey Curtis <jcurtis2@illinois.edu>

zdaq12 committed:
Update src/scenario.F90

Co-authored-by: Jeffrey Curtis <jcurtis2@illinois.edu>

zdaq12 committed:
Update src/scenario.F90

Co-authored-by: Jeffrey Curtis <jcurtis2@illinois.edu>

zdaq12 committed:
formatting

zdaq12 committed:
conflicts

Removed redundant lines and cleaned up the integration logic for dry deposition loss rate calculation.

zdaq12 committed:
improve comments

jcurtis2 commented on test/loss/test_loss_14.sh:
```diff
@@ -0,0 +1,18 @@
+
+../../numeric_diff --by col --rel-tol 1e-8 \
+		loss_modal_drydep_aero_time_ref.txt \
+		out/loss_modal_drydep_aero_time.txt
```
Should this reference file be `loss_modal_drydep_quadpack_aero_time_ref.txt‎` ? The two reference files results are quite different (relative to the specified testing relative tolerances).

zdaq12 replied:
Yes! Fixed that and some missing newlines

zdaq12 committed:
fix incorrect reference file name

zdaq12 committed:
Fix missing newline at end of test_loss_14.sh

zdaq12 committed:
Fix newline at end of test_loss_13.sh
